velox/substrait/SubstraitToVeloxPlanValidator.cpp

```diff
@@ -0,0 +1,336 @@
+            break;
+          default:
+            VLOG(0) << "Field is Expected.";
+            return false;
+        }
+      }
+    } else if (scalarFunction.args().size() == 2) {
+      // Expect there being two args. One is field and the other is literal.
+      bool fieldExists = false;
+      bool litExists = false;
+      for (const auto& param : scalarFunction.args()) {
+        auto typeCase = param.rex_type_case();
+        switch (typeCase) {
+          case ::substrait::Expression::RexTypeCase::kSelection: {
+            fieldExists = true;
+            break;
+          }
+          case ::substrait::Expression::RexTypeCase::kLiteral: {
+            litExists = true;
+            break;
+          }
+          default:
+            VLOG(0) << "Type case: " << typeCase
+                    << " is not supported in ReadRel.";
+            return false;
+        }
+      }
+      if (!fieldExists || !litExists) {
+        VLOG(0) << "Only the case of Field and Literal is supported.";
+        return false;
+      }
+    } else {
+      VLOG(0) << "More than two args is not supported in ReadRel.";
+      return false;
+    }
+  }
+  std::unordered_set<std::string> supportedFilters = {
+      "is_not_null", "gte", "gt", "lte", "lt"};
+  std::unordered_set<std::string> supportedTypes = {"opt", "req", "fp64"};
+  for (const auto& funcSpec : funcSpecs) {
+    // Validate the functions.
+    auto funcName = subParser_->getSubFunctionName(funcSpec);
+    if (supportedFilters.find(funcName) == supportedFilters.end()) {
+      VLOG(0) << "Validation failed due to " << funcName
+              << " was not supported in ReadRel.";
+      return false;
+    }
+
+    // Validate the types.
+    std::vector<std::string> funcTypes;
+    subParser_->getSubFunctionTypes(funcSpec, funcTypes);
+    for (const auto& funcType : funcTypes) {
+      if (supportedTypes.find(funcType) == supportedTypes.end()) {
+        VLOG(0) << "Validation failed due to " << funcType
+                << " was not supported in ReadRel.";
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+bool SubstraitToVeloxPlanValidator::validate(
+    const ::substrait::Rel& substraitRel) {
+  if (substraitRel.has_aggregate()) {
+    return validate(substraitRel.aggregate());
+  }
+  if (substraitRel.has_project()) {
+    return validate(substraitRel.project());
+  }
+  if (substraitRel.has_filter()) {
+    return validate(substraitRel.filter());
+  }
+  if (substraitRel.has_read()) {
+    return validate(substraitRel.read());
+  }
+  return false;
+}
+
+bool SubstraitToVeloxPlanValidator::validate(const ::substrait::RelRoot& root) {
+  if (root.has_input()) {
+    const auto& substraitRel = root.input();
+    return validate(substraitRel);
+  }
+  return false;
+}
+
+bool SubstraitToVeloxPlanValidator::validate(
+    const ::substrait::Plan& substraitPlan) {
+  functions::prestosql::registerAllScalarFunctions();
```

Comment:
Substrait-to-Velox conversion is a generic logic that could work with Presto or Spark or any other function package. It would be nice not to hard-code Presto functions here and rather expect the caller to register the functions they'd like to use.